### PR TITLE
ROX-10665: required attributes auth provider

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

Within Managed Services we want to set required attributes returned by auth providers, so that we can ensure only users from the same Red Hat organization are allowed to login. Other Red Hat users shouldn't be issued a Rox token.

For the solution itself, initially we wanted to opt into a non-API exposed implementation, but since we want to support customers changing the auth provider, which in itself triggers a re-creation of it, I had to add additional fields to `storage.AuthProvider`.
The PR does not _yet_ allow for required attributes for _all_ auth providers, but instead limits modifications to required attributes. I'm happy to less strict if we decide its fine for us.

Right now, the following has been done as a summary:
- Create a new field `RequiredAttributes` on the `storage.AuthProvider`.
- Create a new interface `AttributeChecker`, which is added to the `Provider` interface.
- Create an implementation of `AttributeChecker` called `RequiredAttributesChecker`, which will take a list of `RequiredAttribute` and verify whether the user has the attributes set correctly.
- Add a `ProviderOption` that will allow us to inject the `AttributeChecker` during `AuthProvider`.
- Disallow setting / modifications to `RequiredAttribute` from API, as we want to limit changes being done only by the injected configuration from within managed services ([here is the ADR that goes into detail which mechanism we will use for this](https://docs.google.com/document/d/1piqcfXd_H9jVuFxd0vaqX4wEo2c3nOMdoEnGRQ4XCGM/edit#heading=h.h06w8h600t3e))

What still has to be done after this PR has been merged:
- Once the ADR and mechanism is implemented to receive the auth provider configuration for RH SSO provider in managed service context, the auth provider needs to be created with `ProviderOption`.
- Wire modification within the UI to make changes to required attributes so we ship required attributes as a new feature for all auth providers.
- Add E2E tests going through the flow of adding an auth provider with required attributes, verifying that login is successful / unsuccessful with matching / non-matching attributes.

Once an auth provider has been created with required attributes, each required attribute will be listed:
Screenshot 2022-05-19 at 13 11 31

If a user fails to provide the required attributes, he will be rejected with the following error message:
Screenshot 2022-05-19 at 13 14 17

Overview of the auth provider page:
Screenshot 2022-05-19 at 13 12 11

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
- Unit tests
- Manual testing:
  - Create auth provider (i.e. openshift auth / auth0) via API, set required attributes (i.e. name).
Sample auth provider via Auth0:
```json
{
    "name": "Test required attributes",
    "type": "auth0",
    "uiEndpoint": "localhost:8000",
    "enabled": true,
    "active": true,
    "config": {
        "client_id": "<redacted>",
        "issuer": "<redacted>",
        "mode": "fragment"
    },
    "requiredAttributes": [
        {
            "attributeKey": "name",
            "attributeValue": "<username>"
        }
    ]
}
```
Sample auth provider for OpenShift Auth:
```json
{
    "name": "Test req attributes",
    "type": "openshift",
    "config": {},
    "uiEndpoint": "localhost:8000",
    "enabled": true,
    "active": true,
    "requiredAttributes": [
        {
            "attributeKey": "name",
            "attributeValue": "haus"
        }
    ]
}
```
  - Make sure to explicitly set the minimum access role, otherwise you could be presented with the error `no valid role, please contact your system administrator`.
  - Ensure users with non-matching attributes will be denied login.
  - Ensure users with matching attributes will be allowed login and successfully assigned the minimum role (if no other access role exists).
  - Ensure active auth providers cannot be changed via PutAuthProviders.

### CODE REVIEW

File: proto/storage/auth_provider.proto
Comment (vjwilson) [inline proto/storage/auth_provider.proto:36]: Sorry if I missed this in the arch discussion, but why are we creating two kinds of "attributes": the existing `config` above, and this "config" with an additional property?

File: central/authprovider/service/service_impl.go
Comment (ivan-degtiarenko) [inline central/authprovider/service/service_impl.go:175]: Seems like `authproviders.WithAttributeChecker(providerReq)` would not do anything

File: central/authprovider/service/service_impl.go
Comment (ivan-degtiarenko) [inline central/authprovider/service/service_impl.go:213]: Seems like `authproviders.WithAttributeChecker(request)` would not do anything

File: pkg/auth/authproviders/identity.go
Comment (ivan-degtiarenko) [inline pkg/auth/authproviders/identity.go:35]: I'm still not sure this is the right place to call `AttributeChecker`. I understand that `CreateRoleBasedIdentity` is the single point of transformation between ` *AuthResponse` and `*v1.AuthStatus`, but method name and body do not really correlate with the fact that this check is here.

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (ivan-degtiarenko) [inline ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:454]: I wonder if we should allow for multiple required attributes instead of one. There is no use for multiple required attributes right now, and UI can be simplified(both code and how it looks)

File: pkg/auth/user/required_attribute_checker_test.go
Comment (ivan-degtiarenko) [inline pkg/auth/user/required_attribute_checker_test.go:60]: description does not correspond with test

File: pkg/auth/user/required_attribute_checker_test.go
Comment (ivan-degtiarenko) [inline pkg/auth/user/required_attribute_checker_test.go:14]: also suggestions for tests:

1. `required attribute set but value does not match`
2. `required attribute in the map but empty array value`

**Review by ivan-degtiarenko — COMMENTED** [review]
Would it be possible to post a full-size screenshot of auth provider page with required attributes in it?

File: pkg/auth/authproviders/identity.go
Comment (rukletsov) [inline pkg/auth/authproviders/identity.go:37]: This checks `Rox` claims, i.e., attributes and their values we add to _our_ tokens while I thought we want to check claims in the IdP's tokens. I understand that the latter is tricky to implement in a uniform way so we should make sure we explain we talk about attributes "known" to us.

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [reply central/authprovider/service/service_impl.go:175]: Because we don't allow required attributes here?

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [inline central/authprovider/service/service_impl.go:200]: no period at the end of error messages

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [reply central/authprovider/service/service_impl.go:197]: Some might be using our API directly.

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [inline central/authprovider/service/service_impl.go:195]: I'm not sure we need this check because changes to active auth providers are not allowed anyway (enforced by `registry.DeleteProvider()` below.

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [inline central/authprovider/service/service_impl.go:169]: I think it's fine to allow setting this field via API. The only problem I see is that users might wonder what this field does and request documentation on it...

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (rukletsov) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:454]: Whatever we do, our UI should be in sync with the backend.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:213]: Not entirely true for the Put case.

Imagine we have the RH SSO auth provider within managed services, and customer's decide to change it within the UI, this would trigger this re-creation. We have the required attributes sent from the FE to BE within the API request, and would need to re-create the auth provider with the attribute checker, since in this case the required attributes would be set -- otherwise, you're right, this would not be required.
If we were to delete this now, upon changing the initial provider, customers will get a different behavior after a Put operation.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:175]: For the create case, I agree we can remove this right here, as in theory we won't allow creation via API with required_attributes set.

File: pkg/auth/authproviders/registry_impl.go
Comment (dhaus67) [reply pkg/auth/authproviders/registry_impl.go:168]: Shouldn't have been pushed, reverted that one.

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (dhaus67) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:29]: Added it once more.

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (dhaus67) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:494]: Moved it to the up.

File: pkg/auth/user/required_attribute_checker.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_checker.go:32]: Don't think thats a case we need to explicitly handle as error case, I've added the error if the values do not match and empty strings are handled via `StringFind`, I've also added a test case for this.

File: pkg/auth/user/required_attribute_checker.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_checker.go:29]: Right, not strictly required, moved to `map[string][]string` instead.

File: pkg/auth/user/required_attribute_checker.go
Comment (stehessel) [reply pkg/auth/user/required_attribute_checker.go:34]: Why is that?

File: pkg/auth/user/required_attribute_checker.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_checker.go:36]: Removed that one, shouldn't have been pushed in the first place.

File: pkg/auth/user/attribute_checker.go
Comment (dhaus67) [reply pkg/auth/user/attribute_checker.go:5]: Out of the choices, I favor `AttributeVerifier` the most. Made the naming changes accordingly.

File: pkg/auth/user/required_attribute_checker.go
Comment (stehessel) [reply pkg/auth/user/required_attribute_checker.go:36]: Shouldn't the error already be logged? Why log again?

File: pkg/auth/user/required_attribute_checker.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_checker.go]: Handled the error separately with two distinct error message + a reference of the missing attribute.
I did not include the value in case the values do not match as its potentially something we do not want to leak to unauthorized users.

File: central/authprovider/service/service_impl.go
Comment (stehessel) [reply central/authprovider/service/service_impl.go:213]: Just wondering, could we check if the current user would lock themselves out of the system? This is in case a new required attributed is added that is not actually provided by the identity provider. Perhaps that would even be the desired behaviour, but seems like a potentially bad UX.

**Comment by stehessel** [comment]
@dhaus67 I want to commend you for a clear and easy to understand PR description! 🚀

File: proto/storage/auth_provider.proto
Comment (dhaus67) [reply proto/storage/auth_provider.proto:29]: After re-reading the comment, I see that it could be misinterpreted. I aligned the comment more with Alex's suggestion.

File: pkg/auth/authproviders/identity.go
Comment (dhaus67) [reply pkg/auth/authproviders/identity.go:35]: Hm, I get your point, similar to the `authResp.Claims == nil` check, its currently doing more than expected.
The issue I see with potentially splitting this into two calls would be if clients, like the `authprovider` service, which currently calls `CreateRoleBasedIdentity` would suddenly also make sure they verify it beforehand. I don't think we should move that responsibility to them.

I don't necessarily have a _better_ name for the function itself, not something that would be of the likes of `VerifyAndCreate..` which is _meh_ 😞

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (dhaus67) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:455]: I don't really see the overhead of having multiple required attributes right now. I also don't see how much code it would save on the FE side, as the benefit would be to not have a map call.
If we anyways later on potentially have the goal of providing this with users, I don't see a big overhead for keeping multiple required ones for now and I'm not sure it would significantly simplify things FE / BE wise.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:195]: While that's true, it could be the case that users add another identity provider and change the RH SSO provider, there's nothing stopping them from doing that. I'd rather be on the defensive side with this one and make sure we cover those cases.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:169]: I agree with the sentiment that that may be the case, but on the top of my head couldn't come up with a different, non-hackish solution to allow this specific setting to be set with auth provider.
If we had separate proto definitions for API presentation and storage, we could handle this better I feel like.

File: proto/storage/auth_provider.proto
Comment (dhaus67) [reply proto/storage/auth_provider.proto:36]: Essentially what Ivan already mentioned: We wanted to avoid name-clashes with other config settings + if we decide to add multiple required attributes, it would be cumbersome to do something like `requiredAttribute1=value1,requiredAttribute2=value2`.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:213]: For making changes via `PutProvider`, the provider that will be updated has to be non-active, so the user wouldn't be _immediately_ locked out but would have the possibility to do a i.e. a test login after adding new settings.

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (dhaus67) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx]: Hm, I changed it as you mentioned, looking at the screenshot it seems like we behave differently for info regarding i.e. minimum access role or provider specific configuration.
Seeing as their respective info alert is below their list / field, should we do the same here?

Screenshot for reference:
Screenshot 2022-05-17 at 04 33 04

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx:455]: I think the important point is how it looks. 

For now, our only use case is org_id check(single attribute) and creating an entirely new section might be a little bit of overkill. 
I would be curious to hear @vjwilson opinion on the topic. Maybe, we can unite this section with the one above or below(and potentially generalize the name)?

File: pkg/auth/user/required_attribute_verifier.go
Comment (ivan-degtiarenko) [inline pkg/auth/user/required_attribute_verifier.go:22]: can we make it pointer receiver function?

File: pkg/auth/authproviders/registry_impl.go
Comment (dhaus67) [reply pkg/auth/authproviders/registry_impl.go]: 😵‍💫 

Now its gone - for sure this time.

File: pkg/auth/user/required_attribute_verifier.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_verifier.go:22]: Why? 
It's not required to be used here as we never make changes (and also shouldn't) to the list of required attributes.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:197]: That's a good point, I've dropped the checks I've done regarding changes within required attributes and used the stored values from the provider during creation.

File: pkg/auth/authproviders/identity.go
Comment (dhaus67) [reply pkg/auth/authproviders/identity.go:36]: I feel like we are jumping a bit around the topics, on one hand I agree that doing it within `CreateRoleBasedIdentity` is not desirable, we should clarify this within this PR what we ultimately decide to do here.

Regarding the additional checks within `ExchangeToken` that may or may not be moved to `registry_httphandler`, I'd vote to address them later on. The PR is already quite big and it would increase scope & complexity.

Coming back to `CreateRoleBasedIdentity` - what do you think, should we keep as is, with the comment as a hint that _more_ is done ? Personally, I'm fine with the additional comment.

File: pkg/auth/authproviders/identity.go
Comment (dhaus67) [reply pkg/auth/authproviders/identity.go]: I get the sentiment, I've tried to address this by making more comments, ensuring with a reference to tokens.RoxClaims, that we only take those into account and not IdP specific claims.

**Review by vjwilson — APPROVED** [review]
Front-end changes OK now.

File: ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/AccessControl/AuthProviders/AuthProviderForm.tsx]: I agree that keeping the design is better than involving UX 👍

File: pkg/auth/user/required_attribute_verifier.go
Comment (ivan-degtiarenko) [reply pkg/auth/user/required_attribute_verifier.go:22]: My understanding is that value receiver would result in an additional copy of `checkRequiredAttributesImpl` instance on every `Verify` call. Could safe some bits

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [reply central/authprovider/service/service_impl.go]: I'm still not entirely convinced that we should not allow creation via API. Avoiding exceptions and special cases is a bliss!

File: pkg/auth/authproviders/identity.go
Comment (rukletsov) [reply pkg/auth/authproviders/identity.go:36]: FWIW, I'm not quite happy with expanding `CreateRoleBasedIdentity()` either but I don't have a better suggestion. Specifically, it's unfortunate that we deny login attempt _after_ the auth provider has given green light. Ideally, we would have a generic `ProcessHTTPRequest()` implementation that calls specific version first and then does the required attribute check. Not sure what's the best way to express it in Golang.

File: pkg/auth/user/required_attribute_checker.go
Comment (rukletsov) [reply pkg/auth/user/required_attribute_checker.go]: These wrappers provide default values and hence ensuring proto3 promise on missing fields. Check for example https://github.com/stackrox/stackrox/blob/ce95accd9438afc842d0ac6df97e15e7910f2d0a/generated/storage/role.pb.go#L141-L183.

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [reply central/authprovider/service/service_impl.go:168]: What I'm saying is that let's mark this required attributes feature experimental and roll it out _normally_, meaning, everyone can use this feature via API.

File: central/authprovider/service/service_impl.go
Comment (rukletsov) [reply central/authprovider/service/service_impl.go]: So what you say is that if for whatever reason we ship with several auth providers so that users can activate the non-RHSSO one, log in via it, and then eliminate the required attribute check effectively letting every RH user to their instance. I didn't think about that, but it sounds like they should better not do that for their own benefit : ).

I'm also not sure how to create not `active` auth providers. If we can ensure our auth provider is `active` and hence can't be changed, we might just rely on that.

File: pkg/auth/user/required_attribute_verifier.go
Comment (rukletsov) [inline pkg/auth/user/required_attribute_verifier.go:30]: Similar to the above, there are no clients and no tokens here. You _happen_ to know that `Verify()` will be called in such context(s) today but this is an incorrect assumption in general. I think all these errors shall be unclassified and the caller will wrap them in `errox.NoCredentials`. You won't be able to check for the error class in tests but all errors yielded here are of the same class so it's okay. (To be pedantic, the only caller of this function does not issue tokens either so you might need to adjust the comment there 😄).

**Review by rukletsov — APPROVED** [review]
I'm going to approve to unblock. There are two things I'm unsure about:
* prohibition to specify required attributes in `POST`
* extra check in `PUT`

Let's catch up offline if you'd like to discuss this.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go:168]: I'm fine with making this available during creation via API.
I'm wondering whether a changelog entry would also be in store - but we can also add it later.

File: pkg/auth/user/required_attribute_verifier.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_verifier.go:22]: Feel like with the small set of array elements we deal with it wouldn't necessarily safe anything, but the mantra is anyways when in doubt use pointer receivers  🤷  -- so lets go ahead and use them!

File: pkg/auth/authproviders/identity.go
Comment (dhaus67) [reply pkg/auth/authproviders/identity.go:36]: @ivan-degtiarenko I see, now I understood better what you mean. I could agree with this for the time being.  I've moved the check within `providersHTTPHandler`.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go]: Clarified below, added the ability to create providers with `required_attributes` set once more.

File: central/authprovider/service/service_impl.go
Comment (dhaus67) [reply central/authprovider/service/service_impl.go]: I checked via API to create a auth provider:
```bash
curl --location --request POST 'https://localhost:8000/v1/authProviders' \
--header 'Authorization: Basic <redacted>' \
--header 'Content-Type: application/json' \
--data-raw '{
    "id": "",
    "name": "Test",
    "type": "auth0",
    "uiEndpoint": "localhost:8000",
    "enabled": true,
    "config": {
        "client_id": "<redacted>",
        "issuer": "<redacted>",
        "mode": "fragment"
    },
        "requiredAttributes": [
        {
            "attributeKey": "name",
            "attributeValue": "haus"
        }
    ]
}'
```

This will leave the created auth provider with `active: false`. Via `PUT` I'm able to make changes.
 _After_ logging in with the specific provider for the first time, the provider will be set active. 
So there would be a window between creation + first login where this could potentially be changed.

For Managed Services a user would have to login with RH SSO provider at least once (which, in turn, would mark it as active, thus never being able to change it). 

So I'm fine with deleting this for now.

File: pkg/auth/user/required_attribute_verifier.go
Comment (ivan-degtiarenko) [inline pkg/auth/user/required_attribute_verifier.go:32]: Potential improvement in the future would be to collect the list of missing attributes here. Can be done in a separate PR though.

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM except for one suggestion for future improvement. 
Special thanks for new tests in `registry_httphandler_test.go`!

File: pkg/auth/user/required_attribute_verifier.go
Comment (dhaus67) [reply pkg/auth/user/required_attribute_verifier.go:32]: I'll address this in a follow-up, I also wanted to refactor the test suite of `registry_httphandler` as there is some room for improvement.

**Pull request merged**
